freeipa/src/main/java/com/sequenceiq/freeipa/flow/freeipa/upscale/action/FreeIpaUpscaleActions.java

```diff
@@ -349,11 +353,28 @@ private boolean isNewInstancesWithoutTlsCert(InstanceMetaData instanceMetaData)
         };
     }
 
-    @Bean(name = "UPSCALE_BOOTSTRAPPING_MACHINES_STATE")
-    public Action<?, ?> bootstrappingMachinesAction() {
+    @Bean(name = "UPSCALE_UPDATE_CLUSTERPROXY_REGISTRATION_PRE_BOOTSTRAP_STATE")
+    public Action<?, ?> updateClusterProxyRegistrationPreBootstrapAction() {
         return new AbstractUpscaleAction<>(StackEvent.class) {
             @Override
             protected void doExecute(StackContext context, StackEvent payload, Map<Object, Object> variables) {
+                Stack stack = context.getStack();
+                stackUpdater.updateStackStatus(stack.getId(), getInProgressStatus(variables), "Update cluster proxy registration pre bootstrap");
+                List<String> instanceIds = getInstanceIds(variables);
+                Set<InstanceMetaData> newInstances = instanceMetaDataService.getByInstanceIds(instanceIds);
+                boolean allNewInstanceHasFqdn = newInstances.stream().allMatch(im -> StringUtils.isNotBlank(im.getDiscoveryFQDN()));
```

Comment:
yes, but if it makes to the handler then it is 2 separate kind of registration could happen. At this time of the flow, if there is any instance without FQDN we don't want any registration to happen. While this should not happen, it's more future proof this way.

Reply:
Just one remark: allMatch does return true even if the stream is empty (I was not aware of it previously). Have you taken that in consideration as well?
I guess it would be a major error if instanceIds is empty and maybe the flow won't make it this long at all, so this might be a theoretical discussion.

Reply:
And as I see that case would be handled in the ClusterProxyRegistrationHandler.

Reply:
yes, I handled the empty case, so it shouldn't be an issue. Also it's covered with tests